src/transformStateWithClones.js

```
@@ -7,7 +8,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let resultState = { ...state };
+  const historyState = [];
+
+  for (const action of actions) {
+    let newState = { ...resultState };
+
+    if (action.type === 'clear') {
+      newState = {};
+    }
+
+    if (action.type === 'addProperties') {
+      newState = { ...newState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
```

Because the checklist requires a `switch` structure, it also requires a `default` case for error handling. Exact checklist text: "3. [CODE STYLE]: switch/case should always have default case for error handling." Add a `default` branch to explicitly handle unexpected `action.type` values (for example, leave the state unchanged or log/throw a clear error) instead of silently relying on the absence of matching branches.